cp.apple.finalcutpro.activeCommandSet not updating when cp.apple.finalcutpro.activeCommandSetPath changes

`cp.apple.finalcutpro.activeCommandSet` should automatically update when `cp.apple.finalcutpro.activeCommandSetPath` changes, as it's being monitored - but it doesn't. However, if you type this in the Debug Console:

```lua
cp.apple.finalcutpro.activeCommandSetPath:watch(function() print "CHANGE!" end)
```

...you get the change notification when `cp.apple.finalcutpro.activeCommandSetPath` changes.

@randomeizer writes:

> Had a quick look at activeCommandSetPath, and believe I know what the issue is, but don't have the brain power to fix it tonight. Basically an internal caching issue.

```lua
--- cp.apple.finalcutpro.activeCommandSetPath <cp.prop: string>
--- Field
--- Gets the 'Active Command Set' value from the Final Cut Pro preferences
function fcp.lazy.prop:activeCommandSetPath()
    return self.preferences:prop("Active Command Set", self:defaultCommandSetPath())
end

--- cp.apple.finalcutpro.activeCommandSet <cp.prop: table; live>
--- Variable
--- Contins the 'Active Command Set' as a `table`. The result is cached, but
--- updated automatically if the command set changes.
function fcp.lazy.prop:activeCommandSet()
    return prop(function()
        local path = self:activeCommandSetPath()
        local commandSet = fcp.commandSet(path)
        ----------------------------------------------------------------------------------------
        -- Reset the command cache since we've loaded a new set:
        ----------------------------------------------------------------------------------------
        self._activeCommands = nil

        return commandSet
    end)
    :cached()
    :monitor(self.activeCommandSetPath)
end
```

```lua
> cp.dev.test("cp.prop")()
2022-04-24 08:40:27: 
2022-04-24 08:40:27: [ERROR] cp.prop > Prop Extend: .../Documents/GitHub/CommandPost/src/tests/cp/prop_test.lua:394: attempt to index a nil value (field 'isMethod')
stack traceback:
	.../Documents/GitHub/CommandPost/src/tests/cp/prop_test.lua:394: in field 'executeFn'
	.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:260: in function <.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:260>
	[C]: in function 'xpcall'
	.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:260: in function <.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:218>
	(...tail calls...)
	.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:412: in local 'runFn'
	.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:393: in method '_run'
	.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:403: in function <.../Documents/GitHub/CommandPost/src/extensions/cp/test.lua:396>
	(...tail calls...)
	[C]: in function 'xpcall'
	...ents/GitHub/CommandPost/src/extensions/hs/_coresetup.lua:546: in function <...ents/GitHub/CommandPost/src/extensions/hs/_coresetup.lua:522>
2022-04-24 08:40:27: 
2022-04-24 08:40:27: [RESULT] cp.prop: passed: 299; failed: 0; errors: 1; time: 0.0134s
```

--- a/src/extensions/cp/apple/commandeditor/init.lua
+++ b/src/extensions/cp/apple/commandeditor/init.lua
@@ -8,19 +8,73 @@ local log               = require "hs.logger".new "cmdEditor"
 
 local fnutils           = require "hs.fnutils"
 
+local fn                = require "cp.fn"
 local shortcut          = require "cp.commands.shortcut"
 local tools             = require "cp.tools"
 
 local contains          = fnutils.contains
 local copy              = fnutils.copy
+local find              = string.find
+local insert            = table.insert
+local lower             = string.lower
 local tableContains     = tools.tableContains
+local sort              = fn.table.sort
 
 local mod = {}
 
 --- cp.apple.commandeditor.padKeys -> table
---- Variable
---- Table of Pad Keys
+--- Constant
+--- List of number keys on the number pad. Also mapped with the key name set to `true` for lookup purposes.
 mod.padKeys = { "*", "+", "/", "-", "=", "0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "clear", "enter" }
+
+-- map each pad key to true for quick lookup
+for _, key in ipairs(mod.padKeys) do
+    mod.padKeys[key] = true
+end
+
+--- cp.apple.commandeditor.supportedModifiers -> table
+--- Constant
+--- The list of supported modifiers
+mod.supportedModifiers = { "shift", "control", "option", "command", "function" }
+
+-- map each modifier to true for quick lookup
+for _, modifier in ipairs(mod.supportedModifiers) do
+    mod.supportedModifiers[modifier] = true
+end
+
+local characterStringKeyCodeMap = {
+    [" "]                           = "space",
+    NSF1FunctionKey                 = "f1",
+    NSF2FunctionKey                 = "f2",
+    NSF3FunctionKey                 = "f3",
+    NSF4FunctionKey                 = "f4",
+    NSF5FunctionKey                 = "f5",
+    NSF6FunctionKey                 = "f6",
+    NSF7FunctionKey                 = "f7",
+    NSF8FunctionKey                 = "f8",
+    NSF9FunctionKey                 = "f9",
+    NSF10FunctionKey                = "f10",
+    NSF11FunctionKey                = "f11",
+    NSF12FunctionKey                = "f12",
+    NSF13FunctionKey                = "f13",
+    NSF14FunctionKey                = "f14",
+    NSF15FunctionKey                = "f15",
+    NSF16FunctionKey                = "f16",
+    NSF17FunctionKey                = "f17",
+    NSF18FunctionKey                = "f18",
+    NSF19FunctionKey                = "f19",
+    NSF20FunctionKey                = "f20",
+    NSUpArrowFunctionKey            = "up",
+    NSDownArrowFunctionKey          = "down",
+    NSLeftArrowFunctionKey          = "left",
+    NSRightArrowFunctionKey         = "right",
+    NSInsertFunctionKey             = "insert",
+    NSDeleteFunctionKey             = "delete",
+    NSHomeFunctionKey               = "home",
+    NSEndFunctionKey                = "end",
+    NSPageUpFunctionKey             = "pageup",
+    NSCarriageReturnCharacter       = "return",
+}
 
 --- cp.apple.commandeditor.characterStringToKeyCode() -> string
 --- Function
@@ -35,43 +89,12 @@ function mod.characterStringToKeyCode(input)
 
     local result = tostring(input)
 
-    if input == " "                                     then result = "space"       end
-    if string.find(input, "NSF1FunctionKey")            then result = "f1"          end
-    if string.find(input, "NSF2FunctionKey")            then result = "f2"          end
-    if string.find(input, "NSF3FunctionKey")            then result = "f3"          end
-    if string.find(input, "NSF4FunctionKey")            then result = "f4"          end
-    if string.find(input, "NSF5FunctionKey")            then result = "f5"          end
-    if string.find(input, "NSF6FunctionKey")            then result = "f6"          end
-    if string.find(input, "NSF7FunctionKey")            then result = "f7"          end
-    if string.find(input, "NSF8FunctionKey")            then result = "f8"          end
-    if string.find(input, "NSF9FunctionKey")            then result = "f9"          end
-    if string.find(input, "NSF10FunctionKey")           then result = "f10"         end
-    if string.find(input, "NSF11FunctionKey")           then result = "f11"         end
-    if string.find(input, "NSF12FunctionKey")           then result = "f12"         end
-    if string.find(input, "NSF13FunctionKey")           then result = "f13"         end
-    if string.find(input, "NSF14FunctionKey")           then result = "f14"         end
-    if string.find(input, "NSF15FunctionKey")           then result = "f15"         end
-    if string.find(input, "NSF16FunctionKey")           then result = "f16"         end
-    if string.find(input, "NSF17FunctionKey")           then result = "f17"         end
-    if string.find(input, "NSF18FunctionKey")           then result = "f18"         end
-    if string.find(input, "NSF19FunctionKey")           then result = "f19"         end
-    if string.find(input, "NSF20FunctionKey")           then result = "f20"         end
-    if string.find(input, "NSUpArrowFunctionKey")       then result = "up"          end
-    if string.find(input, "NSDownArrowFunctionKey")     then result = "down"        end
-    if string.find(input, "NSLeftArrowFunctionKey")     then result = "left"        end
-    if string.find(input, "NSRightArrowFunctionKey")    then result = "right"       end
-    if string.find(input, "NSDeleteFunctionKey")        then result = "delete"      end
-    if string.find(input, "NSHomeFunctionKey")          then result = "home"        end
-    if string.find(input, "NSEndFunctionKey")           then result = "end"         end
-    if string.find(input, "NSPageUpFunctionKey")        then result = "pageup"      end
-    if string.find(input, "NSPageDownFunctionKey")      then result = "pagedown"    end
-    if string.find(input, "NSDeleteCharacter")          then result = "delete"      end
-    if string.find(input, "NSCarriageReturnCharacter")  then result = "return"      end
+    input = characterStringKeyCodeMap[input] or input
 
     --------------------------------------------------------------------------------
     -- Convert to lowercase:
     --------------------------------------------------------------------------------
-    result = string.lower(result)
+    result = lower(result)
     return result
 
 end
@@ -87,8 +110,8 @@ end
 ---  * string or nil
 function mod.keypadCharacterToKeyCode(input)
     local result = input
-    for i=1, #mod.padKeys do
-        if input == mod.padKeys[i] then result = "pad" .. input end
+    if mod.padKeys[input] then
+        result = "pad" .. input
     end
     return mod.characterStringToKeyCode(result)
 end
@@ -104,10 +127,10 @@ end
 ---  * table
 function mod.translateModifiers(input)
     local result = {}
-    if string.find(input, "command") then result[#result + 1] = "command" end
-    if string.find(input, "control") then result[#result + 1] = "control" end
-    if string.find(input, "option") then result[#result + 1] = "option" end
-    if string.find(input, "shift") then result[#result + 1] = "shift" end
+    if find(input, "command") then insert(result, "command") end
+    if find(input, "control") then insert(result, "control") end
+    if find(input, "option") then insert(result, "option") end
+    if find(input, "shift") then insert(result, "shift") end
     return result
 end
 
@@ -125,14 +148,17 @@ end
 --- Notes:
 ---  * This function only takes into account 'ctrl', 'alt', 'cmd', 'shift'.
 function mod.modifierMatch(inputA, inputB)
-    local match = true
+    if #inputA ~= #inputB then
+        return false
+    end
 
-    if contains(inputA, "ctrl") and not contains(inputB, "ctrl") then match = false end
-    if contains(inputA, "alt") and not contains(inputB, "alt") then match = false end
-    if contains(inputA, "cmd") and not contains(inputB, "cmd") then match = false end
-    if contains(inputA, "shift") and not contains(inputB, "shift") then match = false end
-
-    return match
+    local sortedA, sortedB = sort(inputA), sort(inputB)
+    for i = 1, #sortedA do
+        if sortedA[i] ~= sortedB[i] then
+            return false
+        end
+    end
+    return true
 end
 
 --- cp.apple.commandeditor.modifierMaskToModifiers() -> table
@@ -160,7 +186,7 @@ function mod.modifierMaskToModifiers(value)
 
     for k, a in pairs(modifiers) do
         if (value & a) == a then
-            table.insert(answer, k)
+            insert(answer, k)
         end
     end
 
@@ -206,7 +232,7 @@ function mod.shortcutsFromCommandSet(id, commandSet)
         local keypadModifier = false
 
         if commmand["modifiers"] ~= nil then
-            if string.find(commmand["modifiers"], "keypad") then keypadModifier = true end
+            if find(commmand["modifiers"], "keypad") then keypadModifier = true end
             modifiers = mod.translateModifiers(commmand["modifiers"])
         elseif commmand["modifierMask"] ~= nil then
             modifiers = mod.modifierMaskToModifiers(commmand["modifierMask"])
@@ -215,18 +241,11 @@ function mod.shortcutsFromCommandSet(id, commandSet)
             end
         end
 
-        if commmand["characterString"] ~= nil then
-            if keypadModifier then
-                keyCode = mod.keypadCharacterToKeyCode(commmand["characterString"])
-            else
-                keyCode = mod.characterStringToKeyCode(commmand["characterString"])
-            end
-        elseif commmand["character"] ~= nil then
-            if keypadModifier then
-                keyCode = mod.keypadCharacterToKeyCode(commmand["character"])
-            else
-                keyCode = mod.characterStringToKeyCode(commmand["character"])
-            end
+        local character = commmand["characterString"] or commmand["character"]
+        if keypadModifier then
+            keyCode = mod.keypadCharacterToKeyCode(character)
+        else
+            keyCode = mod.characterStringToKeyCode(character)
         end
 
         if keyCode ~= nil and keyCode ~= "" then
@@ -238,15 +257,15 @@ function mod.shortcutsFromCommandSet(id, commandSet)
             local cleanedModifiers = {}
             if keypadModifier then
                 for _, v in pairs(modifiers) do
-                    if v == "shift" or v == "control" or v == "option" or v == "command" or v == "function" then
-                        table.insert(cleanedModifiers, v)
+                    if mod.supportedModifiers[v] then
+                        insert(cleanedModifiers, v)
                     end
                 end
             else
                 cleanedModifiers = copy(modifiers)
             end
 
-            shortcuts[#shortcuts + 1] = shortcut.new(cleanedModifiers, keyCode)
+            insert(shortcuts, shortcut.new(cleanedModifiers, keyCode))
         end
     end
 

--- a/src/extensions/cp/apple/commandeditor/init.lua
+++ b/src/extensions/cp/apple/commandeditor/init.lua
@@ -86,17 +86,8 @@ local characterStringKeyCodeMap = {
 --- Returns:
 ---  * Keycode as String or ""
 function mod.characterStringToKeyCode(input)
-
     local result = tostring(input)
-
-    input = characterStringKeyCodeMap[input] or input
-
-    --------------------------------------------------------------------------------
-    -- Convert to lowercase:
-    --------------------------------------------------------------------------------
-    result = lower(result)
-    return result
-
+    return lower(characterStringKeyCodeMap[result] or result)
 end
 
 --- cp.apple.commandeditor.keypadCharacterToKeyCode() -> string

--- a/src/extensions/cp/apple/finalcutpro/init.lua
+++ b/src/extensions/cp/apple/finalcutpro/init.lua
@@ -132,7 +132,7 @@ local doesDirectoryExist                        = tools.doesDirectoryExist
 local stringToHexString                         = tools.stringToHexString
 
 local childMatching                             = axutils.childMatching
-local execute                                   = _G.hs.execute
+local execute                                   = _G["hs"].execute
 local insert                                    = table.insert
 
 -- Load the menu helpers:
@@ -1032,6 +1032,17 @@ function fcp.lazy.prop:activeCommandSet()
     :monitor(self.activeCommandSetPath)
 end
 
+-- cp.apple.finalcutpro._commandShortcuts <table>
+-- Field
+-- Contains the cache of shortcuts that have been retrieved.
+function fcp.lazy.value:_commandShortcuts()
+    -- watch the activeCommandSet and reset the cache when it changes:
+    self.activeCommandSet:watch(function()
+        self._commandShortcuts = {}
+    end)
+    return {}
+end
+
 --- cp.apple.finalcutpro.getCommandShortcuts(id) -> table of hs.commands.shortcut
 --- Method
 --- Finds a shortcut from the Active Command Set with the specified ID and returns a table
@@ -1047,21 +1058,14 @@ function fcp:getCommandShortcuts(id)
         log.ef("ID is required for cp.apple.finalcutpro.getCommandShortcuts.")
         return nil
     end
-    local activeCommands = self._activeCommands or {}
-    local shortcuts = activeCommands[id]
+    local shortcuts = self._commandShortcuts[id]
     if not shortcuts then
         local commandSet = self:activeCommandSet()
         shortcuts = commandeditor.shortcutsFromCommandSet(id, commandSet)
         if not shortcuts then
             return nil
         end
-        ----------------------------------------------------------------------------------------
-        -- Cache the value for faster access next time:
-        ----------------------------------------------------------------------------------------
-        if not self._activeCommands then
-            self._activeCommands = {}
-        end
-        self._activeCommands[id] = shortcuts
+        self._commandShortcuts[id] = shortcuts
     end
     return shortcuts
 end

--- a/src/extensions/cp/fn.lua
+++ b/src/extensions/cp/fn.lua
@@ -199,11 +199,11 @@ end
 ---
 --- Notes:
 ---  * The comparators are called in the order they are provided.
----  * If no comparators are provided, this will return `nil`.
+---  * If no comparators are provided, returns a `nil` function, which is generally sorted with the standard `<` operator.
 function mod.compare(...)
     local comparators = packArgs(...)
     if #comparators == 0 then
-        return
+        return -- generally defaults to the standard `a < b` operator
     elseif #comparators == 1 then
         return comparators[1]
     end
@@ -306,7 +306,7 @@ end
 
 --- cp.fn.flip(fn) -> function(...) -> function(...) -> any
 --- Function
---- A cobinator that flips the order of the next two arguments to a curried function.
+--- A combinator that flips the order of the next two arguments to a curried function.
 ---
 --- Parameters:
 ---  * fn - The function to flip.
@@ -314,6 +314,9 @@ end
 --- Returns:
 ---  * A function that accepts the second argument and returns
 ---     a function expecting the first argument.
+---
+--- Notes:
+---  * If multiple arguments are provided for either function, the order of the arguments within that list are not flipped.
 function mod.flip(fn)
     return function(...)
         local args = pack(...)
@@ -395,10 +398,11 @@ function mod.prefix(fn, ...)
         for i = 1, argCount do
             inputs[i] = args[i]
         end
-        for i = 1, select("#", ...) do
+        local argCount2 = select("#", ...)
+        for i = 1, argCount2 do
             inputs[i + argCount] = select(i, ...)
         end
-        return fn(unpack(inputs))
+        return fn(unpack(inputs, 1, argCount + argCount2))
     end
 end
 

--- a/src/extensions/cp/fn/table.lua
+++ b/src/extensions/cp/fn/table.lua
@@ -333,7 +333,7 @@ end
 
 --- cp.fn.table.sort(...) -> function(table) -> table
 --- Function
---- A combinator that returns a function that accepts a table and sorts it based on the `compareFn`.
+--- A combinator that returns a function that accepts a table and returns a new table, sorted with the compare functions.
 ---
 --- Parameters:
 ---  * ... - The list of compare functions to use, in order.

